### PR TITLE
build: add staticcheck

### PR DESCRIPTION
**Pull request by pmalek**
**What this PR does / why we need it**:

Add https://github.com/dominikh/go-tools/tree/master/cmd/staticcheck to the set of linters we run.

We add this separately from golangci-lint because:

> The author of staticcheck doesn't support or approve the use of staticcheck as a library inside golangci-lint.

### CODE REVIEW

File: docs/api-reference.md [hunk 1]
```diff
@@ -201,7 +201,7 @@ _Appears in:_
 
 _Underlying type:_ `string`
 
-
+KongProtocol is a valid Kong protocol This alias is necessary to deal with https://github.com/kubernetes-sigs/controller-tools/issues/342
```
Comment (pmalek) [inline docs/api-reference.md:204]: @czeslavo This is correct, isn't it? Or was it meant to be hidden?

Comment (czeslavo) [reply docs/api-reference.md:204]: I think so, don't understand why it was ignored before 🤔

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -37,6 +37,12 @@ jobs:
     - name: Run lint
       run: make lint
 
+    - name: Run staticcheck action
+      uses: dominikh/staticcheck-action@v1.3.0
+      with:
+        build-tags: e2e_tests,integration_tests,istio_tests,conformance_tests
+        install-go: false
```
Comment (czeslavo) [inline .github/workflows/test.yaml:44]: Would it work with `run: make staticcheck`? That way we could avoid duplication of the build tags passed to the command.

Comment (pmalek) [reply .github/workflows/test.yaml:44]: We can do this but with the action we get the PR annotations (using Github Actions matchers: https://github.com/dominikh/staticcheck-action/blob/ba605356b4b29a60e87ab9404b712f3461e566dc/action.yaml#L103, ref: https://fusectore.dev/2021/11/19/github-action-problem-matchers.html)

Comment (czeslavo) [reply .github/workflows/test.yaml:44]: Hmm, I think I'd prefer to not have the annotations, but be able to run the very same linting command locally, to be honest. BTW `make lint` target could depend on `staticcheck` in such a case, to not make the staticcheck a special citizen here.

Comment (pmalek) [reply .github/workflows/test.yaml]: Fair enough. Removed this and placed `staticcheck` as a dependency for `lint` Makefile target.

**Review by czeslavo — APPROVED** [review]

**Pull request merged**